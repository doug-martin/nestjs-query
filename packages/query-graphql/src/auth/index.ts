export * from './authorizer';
export * from './default-crud.authorizer';
export * from './tokens';
