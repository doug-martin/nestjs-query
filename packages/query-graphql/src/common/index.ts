export { DTONamesOpts, getDTONames, DTONames, getDTOIdTypeOrDefault } from './dto.utils';
export * from './external.utils';
export * from './resolver.utils';
export * from './object.utils';
