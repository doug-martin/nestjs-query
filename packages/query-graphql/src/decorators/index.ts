export {
  FilterableField,
  FilterableFieldOptions,
  FilterableFieldDescriptor,
  getFilterableFields,
} from './filterable-field.decorator';
export { ResolverMethodOpts } from './resolver-method.decorator';
export {
  CursorConnection,
  FilterableCursorConnection,
  OffsetConnection,
  FilterableOffsetConnection,
  UnPagedRelation,
  FilterableUnPagedRelation,
  Relation,
  FilterableRelation,
  RelationDecoratorOpts,
  RelationTypeFunc,
  getRelations,
} from './relation.decorator';
export * from './resolver-mutation.decorator';
export * from './resolver-query.decorator';
export * from './resolver-field.decorator';
export { Reference, ReferenceDecoratorOpts, ReferenceTypeFunc } from './reference.decorator';
export { ResolverSubscription, SubscriptionResolverMethodOpts } from './resolver-subscription.decorator';
export { InjectPubSub } from './inject-pub-sub.decorator';
export * from './skip-if.decorator';
export * from './aggregate-query-param.decorator';
export * from './hook.decorator';
export * from './decorator.utils';
export * from './hook-args.decorator';
export * from './authorizer.decorator';
export * from './inject-authorizer.decorator';
export * from './inject-custom-authorizer.decorator';
export * from './key-set.decorator';
export * from './authorize-filter.decorator';
export * from './query-options.decorator';
export * from './id-field.decorator';
