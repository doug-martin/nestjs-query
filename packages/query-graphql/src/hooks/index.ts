export * from './hooks';
export { createDefaultHook } from './default.hook';
export * from './tokens';
export * from './types';
