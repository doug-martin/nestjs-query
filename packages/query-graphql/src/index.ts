export * from './types';
export {
  FilterableField,
  FilterableFieldOptions,
  ResolverMethodOpts,
  Relation,
  FilterableRelation,
  CursorConnection,
  FilterableCursorConnection,
  OffsetConnection,
  FilterableOffsetConnection,
  UnPagedRelation,
  FilterableUnPagedRelation,
  RelationTypeFunc,
  RelationDecoratorOpts,
  Reference,
  ReferenceTypeFunc,
  ReferenceDecoratorOpts,
  InjectPubSub,
  BeforeCreateOne,
  BeforeCreateMany,
  BeforeUpdateOne,
  BeforeUpdateMany,
  BeforeDeleteOne,
  BeforeDeleteMany,
  BeforeQueryMany,
  BeforeFindOne,
  InjectAuthorizer,
  InjectCustomAuthorizer,
  Authorize,
  AuthorizerFilter,
  RelationAuthorizerFilter,
  HookArgs,
  MutationHookArgs,
  KeySet,
  QueryOptions,
  IDField,
  IDFieldOptions,
} from './decorators';
export * from './resolvers';
export * from './federation';
export { DTONamesOpts } from './common';
export { NestjsQueryGraphQLModule } from './module';
export { AutoResolverOpts } from './providers';
export { pubSubToken, GraphQLPubSub } from './subscription';
export { Authorizer, CustomAuthorizer, AuthorizerOptions, AuthorizationContext, OperationGroup } from './auth';
export {
  Hook,
  HookTypes,
  BeforeCreateOneHook,
  BeforeCreateManyHook,
  BeforeUpdateOneHook,
  BeforeUpdateManyHook,
  BeforeDeleteOneHook,
  BeforeDeleteManyHook,
  BeforeQueryManyHook,
  BeforeFindOneHook,
} from './hooks';
export { AuthorizerInterceptor, AuthorizerContext, HookInterceptor, HookContext } from './interceptors';
