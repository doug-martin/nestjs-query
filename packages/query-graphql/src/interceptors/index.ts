export * from './hook.interceptor';
export * from './authorizer.interceptor';
