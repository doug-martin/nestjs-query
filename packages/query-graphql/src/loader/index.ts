export * from './find-relations.loader';
export * from './query-relations.loader';
export * from './count-relations.loader';
export * from './aggregate-relations.loader';
export * from './dataloader.factory';
