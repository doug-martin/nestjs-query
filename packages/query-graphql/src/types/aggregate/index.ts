export { AggregateResponseType, AggregateResponseOpts } from './aggregate-response.type';
export { AggregateArgsType } from './aggregate-args.type';
