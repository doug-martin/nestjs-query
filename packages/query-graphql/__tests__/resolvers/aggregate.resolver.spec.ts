import { AggregateQuery, AggregateResponse } from '@nestjs-query/core';
import { Query, Resolver } from '@nestjs/graphql';
import { when, deepEqual, objectContaining } from 'ts-mockito';
import { AggregateArgsType } from '../../src';
import { AggregateResolver, AggregateResolverOpts } from '../../src/resolvers/aggregate.resolver';
import { generateSchema, createResolverFromNest, TestResolverDTO, TestService } from '../__fixtures__';

describe('AggregateResolver', () => {
  const expectResolverSDL = async (opts?: AggregateResolverOpts) => {
    @Resolver(() => TestResolverDTO)
    class TestSDLResolver extends AggregateResolver(TestResolverDTO, opts) {
      @Query(() => TestResolverDTO)
      test(): TestResolverDTO {
        return { id: '1', stringField: 'foo' };
      }
    }
    const schema = await generateSchema([TestSDLResolver]);
    expect(schema).toMatchSnapshot();
  };

  it('should create a AggregateResolver for the DTO', () => expectResolverSDL({ enabled: true }));

  it('should not expose read methods if not enabled', () => expectResolverSDL());

  describe('#aggregate', () => {
    @Resolver(() => TestResolverDTO)
    class TestResolver extends AggregateResolver(TestResolverDTO, { enabled: true }) {
      constructor(service: TestService) {
        super(service);
      }
    }
    it('should call the service query with the provided input', async () => {
      const { resolver, mockService } = await createResolverFromNest(TestResolver);
      const input: AggregateArgsType<TestResolverDTO> = {
        filter: {
          stringField: { eq: 'foo' },
        },
      };
      const aggregateQuery: AggregateQuery<TestResolverDTO> = { count: ['id'] };
      const output: AggregateResponse<TestResolverDTO>[] = [
        {
          count: { id: 10 },
        },
      ];
      when(mockService.aggregate(objectContaining(input.filter!), deepEqual(aggregateQuery))).thenResolve(output);
      const result = await resolver.aggregate(input, aggregateQuery);
      return expect(result).toEqual(output);
    });
  });
});
