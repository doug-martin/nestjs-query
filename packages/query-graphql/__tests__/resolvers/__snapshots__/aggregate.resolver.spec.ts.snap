// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AggregateResolver should create a AggregateResolver for the DTO 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type TestResolverDTOAggregateGroupBy {
  id: ID
  stringField: String
}

type TestResolverDTOCountAggregate {
  id: Int
  stringField: Int
}

type TestResolverDTOMinAggregate {
  id: ID
  stringField: String
}

type TestResolverDTOMaxAggregate {
  id: ID
  stringField: String
}

type TestResolverDTOAggregateResponse {
  groupBy: TestResolverDTOAggregateGroupBy
  count: TestResolverDTOCountAggregate
  min: TestResolverDTOMinAggregate
  max: TestResolverDTOMaxAggregate
}

type Query {
  testResolverDTOAggregate(
    """Filter to find records to aggregate on"""
    filter: TestResolverDTOAggregateFilter
  ): [TestResolverDTOAggregateResponse!]!
  test: TestResolverDTO!
}

input TestResolverDTOAggregateFilter {
  and: [TestResolverDTOAggregateFilter!]
  or: [TestResolverDTOAggregateFilter!]
  id: IDFilterComparison
  stringField: StringFieldComparison
}

input IDFilterComparison {
  is: Boolean
  isNot: Boolean
  eq: ID
  neq: ID
  gt: ID
  gte: ID
  lt: ID
  lte: ID
  like: ID
  notLike: ID
  iLike: ID
  notILike: ID
  in: [ID!]
  notIn: [ID!]
}

input StringFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: String
  neq: String
  gt: String
  gte: String
  lt: String
  lte: String
  like: String
  notLike: String
  iLike: String
  notILike: String
  in: [String!]
  notIn: [String!]
}

`;

exports[`AggregateResolver should not expose read methods if not enabled 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type TestResolverDTOAggregateGroupBy {
  id: ID
  stringField: String
}

type TestResolverDTOCountAggregate {
  id: Int
  stringField: Int
}

type TestResolverDTOMinAggregate {
  id: ID
  stringField: String
}

type TestResolverDTOMaxAggregate {
  id: ID
  stringField: String
}

type Query {
  test: TestResolverDTO!
}

`;
